[[ LifecycleListeners ]] Allow lcb widgets to register and listen for Android lifecycle events

livecodemichael opened the pull request:
This PR adds handlers to the android-utils module allowing lcb widgets to register for Android lifecycle events (at present these are pause and resume).

Note that we wrap the listener object in LiveCodeActivity.java, allowing for storage in a hash set. A possible alternative would be to use an index based container, with register returning the index of the listener in the container and unregister taking the index of the listener to unregister.

livecodemichael committed:
[[ LifecycleListeners ]] Add ability to register for lifecycle events on Android

This patch adds methods to the Android `Engine` class to enable subscription to
the `onPause` and `onResume` lifecycle events. Subscription is done by providing
an object implementing the `LifecycleListener` interface. The ability is designed
to be used internally, and to be wrapped by LCB utility methods.

livecodemichael committed:
[[ LifecycleListeners ]] Add lifecycle listener registration handlers to android utils module

This patch add handlers to the android utils module allowing for
lcb widgets to register to receive lifecycle events.

runrevmark commented:
So I think using an indexed-container would be better. This means we can guarantee an order on calling of OnPause / OnResume. We OnPause in order of addition, and OnResume in reverse order.

There's no need to use/return an index - as the Listeners are objects - unless I'm forgetting something about Java that means we can just use equality?

We can make things super simple - Register appends the given listener object to the list, Unregister removes the most recent matching listener from the list. (By list I mean array which is always dense - so if you delete element 1, all elements above get shuffled back down).

I also think that the code to register/unregister and the list itself should be maintained in the engine view, not the Activity (the Activity is purposely very lightweight).

Finally, I think all registered listeners should pause before the main view does, and resume afterwards - otherwise things they may depend on engine-side might not be in the right state.

(One other thing is that I don't think we need to worry about synchronization, engine/LCB always runs co-operatively so these shouldn't ever be called from two threads at once).

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -3994,7 +4014,32 @@ public int getSystemAppearance()
+		int t_listener_hash_code = System.identityHashCode(p_listener);
+		for (int i = m_lifecycle_listeners.size() - 1; i >= 0; i--)
+		{
+			if (System.identityHashCode(m_lifecycle_listeners.get(i)) == t_listener_hash_code)
```
Can you not just do `m_lifecycle_listeners.get(i) == p_listener`?

runrevmark commented on extensions/modules/android-utils/android-utils.lcb:
```diff
@@ -232,4 +232,109 @@ public handler AndroidHasPermission(in pPermission as String) returns Boolean
+        put nothing into mLifecycleListener
+    end handler
+
+    public handler OnPause() returns nothing
```
I suggest that we recommend people *do not* use public `OnPause` and `OnResume` as the names - and instead use more specific naming using private handlers. This is so we can (in the future) add automatic registration of such lifecycle handlers (i.e. have widget events which do the same thing without having to register / deregister).

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -3368,6 +3378,11 @@ public void onAppLaunched()
 
 	public void onPause()
 	{
+		for (int i = 0; i < m_lifecycle_listeners.size(); i++)
```
Okay so I think we should pause in reverse order. If two components are not related then it doesn't matter.If component B depends on component A, then component A will have been initialized before component B. Therefore A can't care about B, but B might care about A, so B should pause before A.

i.e. Pause in reverse order of registration.

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -3443,6 +3458,11 @@ public void onResume()
 		if (m_wake_on_event)
 			doProcess(false);
+
+		for (int i = m_lifecycle_listeners.size() - 1; i >= 0; i--)
```
This should happen before we `doProcess()` above I think - everything before that causes posts to the event queue, but that call could cause script to run which means all components should have been resumed.

Also this should loop from 0 to size so we resume earlier registered components before later ones.

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -3368,6 +3378,11 @@ public void onAppLaunched()
 
 	public void onPause()
 	{
+		for (int i = m_lifecycle_listeners.size() - 1; i >= 0; i--)
```
We should comment this to explain the ordering:
```
/* Pause registered listeners in reverse order of registration as
 * then components will be paused before any components they
 * depend on. */
```

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -3440,7 +3455,12 @@ public void onResume()
 		dispatchNotifications();
-		
+
+		for (int i = 0; i < m_lifecycle_listeners.size(); i++)
```
```
/* Resume registered listeners in order of registration as then
 * components will be resumed before any components which depend
 * on them. */
```

runrevmark reviewed (changes requested):
Just one last thing :)

The commit message is now stale as it refers to Android activity:
```
[[ LifecycleListeners ]] Add ability to register for lifecycle events on Android

This patch adds methods to the Android `Engine` class to enable subscription to
the `onPause` and `onResume` lifecycle events. Subscription is done by providing
an object implementing the `LifecycleListener` interface. The ability is designed
to be used internally, and to be wrapped by LCB utility methods.
```

runrevmark commented:
@livecode-vulcan review ok 45aef32

livecode-vulcan commented:
💙 review by @runrevmark ok 45aef32aab9203708319a2ff0bdcd9a97fdacaed

livecode-vulcan commented:
😞 test failure 45aef32aab9203708319a2ff0bdcd9a97fdacaed

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/788)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/197)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/197)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/197)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/197)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2314)
* try-community-universal-ios-iphoneos13.4: [pending](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.4/builds/47)
* try-community-universal-ios-iphonesimulator13.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.4/builds/48)
* try-community-universal-mac-macosx10.9: [pending](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2342)
* try-community-x86-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2330)
* try-community-x86_64-linux-debian8: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2329)
* try-community-x86-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2364)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2304)

livecodemichael committed:
[[ docs ]] Correctly parse handler type definitions

The patch updates the docs parser to ensure that it correctly
handles handler type definitions. Previously, it was parsing them
as handler definitions, expecting a corresponding end handler and
a docs entry, causing the parser to get confused.

runrevmark commented:
@livecode-vulcan review ok 8cd3d53

livecode-vulcan commented:
💙 review by @runrevmark ok 8cd3d535fa7ecd98960fd3e64bc4854911fe67d7

livecode-vulcan commented:
😎 test success 8cd3d535fa7ecd98960fd3e64bc4854911fe67d7

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/791)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/200)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/200)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/200)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/200)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2317)
* try-community-universal-ios-iphoneos13.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.4/builds/50)
* try-community-universal-ios-iphonesimulator13.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.4/builds/51)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2345)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2333)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2332)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2367)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2307)
